pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,503 @@
+				return nil, err
+			}
+			return nil, nil
+		}
+		routePathBulkMessageMap := make(map[string]pubsubBulkSubscribedMessage)
+		bulkResponses := make([]pubsub.BulkSubscribeResponseEntry, len(msg.Entries))
+		entryIDIndexMap := make(map[string]int)
+		hasAnyError := false
+		for i, message := range msg.Entries {
+			if entryIDErr := validateEntryID(message.EntryID, i); entryIDErr != nil {
+				bulkResponses[i].Error = entryIDErr
+				hasAnyError = true
+				continue
+			}
+			entryIDIndexMap[message.EntryID] = i
+			if rawPayload {
+				rPath, routeErr := a.getRouteIfProcessable(ctx, route, &(msg.Entries[i]), i, &bulkResponses, string(message.Event), psName, topic)
+				if routeErr != nil {
+					hasAnyError = true
+					continue
+				}
+				if rPath == "" {
+					continue
+				}
+				dataB64 := base64.StdEncoding.EncodeToString(message.Event)
+				var contenttype string
+				if message.ContentType != "" {
+					contenttype = message.ContentType
+				} else {
+					contenttype = "application/octet-stream"
+				}
+				populateBulkSubcribedMessage(&(msg.Entries[i]), dataB64, &routePathBulkMessageMap, rPath, i, msg, false, psName, contenttype)
+			} else {
+				var cloudEvent map[string]interface{}
+				err = json.Unmarshal(message.Event, &cloudEvent)
+				if err != nil {
+					log.Errorf("error deserializing one of the messages in bulk cloud event in pubsub %s and topic %s: %s", psName, msg.Topic, err)
+					bulkResponses[i].Error = err
+					bulkResponses[i].EntryID = message.EntryID
+					hasAnyError = true
+					continue
+				}
+				if pubsub.HasExpired(cloudEvent) {
+					log.Warnf("dropping expired pub/sub event %v as of %v", cloudEvent[pubsub.IDField], cloudEvent[pubsub.ExpirationField])
+					diag.DefaultComponentMonitoring.BulkPubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), topic, 0)
+					if route.deadLetterTopic != "" {
+						_ = a.sendToDeadLetter(psName, &pubsub.NewMessage{
+							Data:        message.Event,
+							Topic:       topic,
+							Metadata:    message.Metadata,
+							ContentType: &message.ContentType,
+						}, route.deadLetterTopic)
+					}
+					bulkResponses[i].EntryID = message.EntryID
+					bulkResponses[i].Error = nil
+					continue
+				}
+				rPath, routeErr := a.getRouteIfProcessable(ctx, route, &(msg.Entries[i]), i, &bulkResponses, cloudEvent, psName, topic)
+				if routeErr != nil {
+					hasAnyError = true
+					continue
+				}
+				if rPath == "" {
+					continue
+				}
+				populateBulkSubcribedMessage(&(msg.Entries[i]), cloudEvent, &routePathBulkMessageMap, rPath, i, msg, true, psName, message.ContentType)
+			}
+		}
+		for path, psm := range routePathBulkMessageMap {
+			invokeErr := a.createEnvelopeAndInvokeSubscriber(ctx, psm, topic, psName, msg, route, &bulkResponses, &entryIDIndexMap, path, policy)
+			if invokeErr != nil {
+				hasAnyError = true
+				err = invokeErr
+			}
+		}
+		if hasAnyError && err != context.Canceled {
```

Comment:
Qq what happens when hasAnyError is true and error is overwritten by any of the following lines since the error occurred and it is context cancelled...then we are not considering that scenario here... Would that lead to messages not being sent to dlq?

Reply:
It is a good question. I thought about it again that in this case, when do I actually need to check for context cancellation. I think it would be good to check for context cancellation, specifically when hearing back errror after invoke calls. So, changing logic a bit here. Please provide your thoughts. I am not individual invokeErrs but overall as want to return back complete bulkResponses, in any case.